lib/ruby_lsp/server.rb

```diff
@@ -1,250 +1,158 @@
+        # If the response include any notifications, go through them and publish each one
+        notifications = result.notifications
+        notifications.each { |n| @writer.write(method: n.message, params: n.params) } if notifications.any?
+
+        if error
+          @writer.write(
+            id: request[:id],
+            error: {
+              code: Constant::ErrorCodes::INTERNAL_ERROR,
+              message: error.inspect,
+              data: request.to_json,
+            },
+          )
+        elsif response != VOID
+          @writer.write(id: request[:id], result: response)
         end
-      end
-    end
 
-    on("textDocument/semanticTokens/full", parallel: true) do |request|
-      store.cache_fetch(request.dig(:params, :textDocument, :uri), :semantic_highlighting) do |document|
-        T.cast(
-          Requests::SemanticHighlighting.new(
-            document,
-            encoder: Requests::Support::SemanticTokenEncoder.new,
-          ).run,
-          LanguageServer::Protocol::Interface::SemanticTokens,
-        )
+        request_time = result.request_time
+        if request_time
+          @writer.write(method: "telemetry/event", params: telemetry_params(request, request_time, error))
+        end
       end
     end
 
-    on("textDocument/semanticTokens/range", parallel: true) do |request|
-      document = store.get(request.dig(:params, :textDocument, :uri))
-      range = request.dig(:params, :range)
-      start_line = range.dig(:start, :line)
-      end_line = range.dig(:end, :line)
-
-      Requests::SemanticHighlighting.new(
-        document,
-        range: start_line..end_line,
-        encoder: Requests::Support::SemanticTokenEncoder.new,
-      ).run
+    sig do
+      params(
+        request: T::Hash[Symbol, T.untyped],
+        request_time: Float,
+        error: T.nilable(Exception),
+      ).returns(T::Hash[Symbol, T.any(String, Float)])
     end
-
-    on("textDocument/formatting", parallel: true) do |request|
+    def telemetry_params(request, request_time, error)
       uri = request.dig(:params, :textDocument, :uri)
+      params = {
+        request: request[:method],
+        lspVersion: RubyLsp::VERSION,
+        requestTime: request_time,
+      }
 
-      Requests::Formatting.new(uri, store.get(uri)).run
-    end.on_error do |error|
-      show_message(Constant::MessageType::ERROR, "Formatting error: #{error.message}")
-    end
-
-    on("textDocument/onTypeFormatting", parallel: true) do |request|
-      uri = request.dig(:params, :textDocument, :uri)
-      position = request.dig(:params, :position)
-      character = request.dig(:params, :ch)
+      if error
+        params[:errorClass] = error.class.name
+        params[:errorMessage] = error.message
 
-      Requests::OnTypeFormatting.new(store.get(uri), position, character).run
-    end
+        log_params = request[:params]&.reject { |k, _| k == :textDocument }
+        params[:params] = log_params.to_json if log_params&.any?
```

Comment:
nitpick: I feel it's clearer if we can avoid lonely operators in 2 consecutive lines with an if statement.

We can keep an if statement but eliminate the lonely operators. WDYT?

```suggestion
        
        log_params = request[:params]
        if log_params
          params[:params] = log_params.reject { |k, _| k == :textDocument }.to_json
        end
```